Prevent tooltip from showing when hovering past the end of script line

larspet opened the pull request:
Fixes #100767

`TextEdit::get_line_column_at_pos()` already had an option to not allow "out of bounds" access, but it only checked if the row was OOB, not the column. This PR renames that argument and adds an extra argument that allows for checking the column as well.

<details>
<summary>Before</summary>

[](https://github.com/user-attachments/assets/5a55d055-fcd8-40b9-867d-80a10da0d17b)
</details>
<details>
<summary>After</summary>

[](https://github.com/user-attachments/assets/b0039ae5-b22a-4bd0-ab53-bf414e8b891a)
</details>

larspet committed:
Prevent tooltip from showing when hovering past end of script line

Mickeon commented:
I wonder, why would one ever want the existing behaviour enough to warrant it being an optional argument?

larspet commented:
> I wonder, why would one ever want the existing behavior enough to warrant it being an optional argument?

For `get_line_column_at_pos()` I can see why it's needed, e.g. in order to put the caret at the last column when clicking past the end of the line. Although I think the argument is backwards, and should be named `p_clamp_position` or something instead.

The reason I added the argument to `get_word_at_pos()` was to give them similar API, and not potentially break current uses of it. That said, looking at the other few existing uses of it in the engine, this new functionality would be wanted there too. I can't speak for any user code already utilizing the function however.

If wanted, I would not be against not adding this new argument, and instead changing the current behavior of `get_word_at_pos()` instead.

dalexeev commented on misc/extension_api_validation/4.3-stable.expected:
```diff
@@ -268,3 +268,10 @@ GH-98441
+---------
+Validate extension JSON: Error: Field 'classes/TextEdit/methods/get_word_at_pos/arguments': size changed value in new API, from 1 to 2.
+
+Added optional argument to get_word_at_pos to disallow positions that are outside of the text boundaries. This matches the arguments for get_line_column_at_pos.
```
```suggestion
Added optional argument to `get_word_at_pos()` to disallow positions that are outside of the text boundaries. This matches the arguments for `get_line_column_at_pos()`. Compatibility method registered.
```

dalexeev commented on scene/gui/code_edit.cpp:
```diff
@@ -443,8 +443,8 @@ void CodeEdit::gui_input(const Ref<InputEvent> &p_gui_input) {
 			if (mm->is_command_or_control_pressed() && mm->get_button_mask().is_empty()) {
-				symbol_lookup_pos = get_line_column_at_pos(mpos);
-				symbol_lookup_new_word = get_word_at_pos(mpos);
+				symbol_lookup_pos = get_line_column_at_pos(mpos, false);
+				symbol_lookup_new_word = get_word_at_pos(mpos, false);
```
Please add `false` for `get_word_at_pos()` here too:

https://github.com/godotengine/godot/blob/367c157beac0b26f386a3d33f9b92f057adf2aeb/scene/gui/code_edit.cpp#L457-L458

dalexeev reviewed (approved):
I'm not a node GUI expert, but the code looks good to me, I tested it and it works fine.

bruvzg commented on scene/gui/text_edit.cpp:
```diff
@@ -4463,6 +4466,11 @@ Point2i TextEdit::get_line_column_at_pos(const Point2i &p_pos, bool p_allow_out_
 		colx -= wrap_indent;
 	}
+
+	if (!p_allow_out_of_bounds && TS->shaped_text_hit_test_grapheme(text_rid, colx) < 0) {
```
Why `shaped_text_hit_test_grapheme` if you can simply compare it to the text size?

```c++
if (!p_allow_out_of_bounds && (colx < 0 || colx > TS->shaped_text_get_size(text_rid).x)) {
```

larspet replied:
That is indeed a simpler alternative.

kitbdev commented:
The behavior of `get_line_column_at_pos` when allow oob is false shouldn't change. A new parameter can be added to it for horizontal checking, and the existing one's name can be changed.
This is because there is code that relies on the current behavior of only checking vertical like in `CodeEdit::get_cursor_shape` for the fold icon, and users may be relying on it too. I would also like to use it for gutters, see #101613.

larspet commented:
I removed the optional argument from `get_word_at_pos` since there really wasn't any point to it. The new behavior follows the expected behavior much more anyways.

Instead, I split the argument in `get_line_column_at_pos` into two (line & column). This way the old behavior is kept (so that e.g. the line-folding `...` icon works), while also optionally allowing to disallow clamping for the column.

kitbdev commented on scene/gui/text_edit.h:
```diff
@@ -658,6 +658,7 @@ class TextEdit : public Control {
 
 #ifndef DISABLE_DEPRECATED
 	void _set_selection_mode_compat_86978(SelectionMode p_mode, int p_line = -1, int p_column = -1, int p_caret = 0);
+	String _get_line_column_at_pos_bind_compat_100913(const Vector2 &p_pos, bool p_allow_out_of_bounds = true) const;
```
```suggestion
	Point2i _get_line_column_at_pos_bind_compat_100913(const Vector2 &p_pos, bool p_allow_out_of_bounds = true) const;
```
Same with the compat file.
I think this should fix the validation error.

larspet replied:
Thanks, changing the return type from `String` was part of it, but I was also missing a `DEFVAL`.

Calinou reviewed (approved):
Tested locally, it works as expected. Code and docs look good to me.

akien-mga commented:
Thanks!
